### ISSUE TITLE
Error trying to compile with GPU support for R

### ISSUE THREAD

**Issue by Hartsord** [dmlc/xgboost#3081]
Hello, I'm new to building packages for R, and I'm getting errors trying to build gpu support for R, as listed here: [http://xgboost.readthedocs.io/en/latest/build.html#installing-r-package-with-gpu-support](http://xgboost.readthedocs.io/en/latest/build.html#installing-r-package-with-gpu-support).

Specifically, I'm getting this error:
```
CMake Error at cmake/modules/FindLibR.cmake:44 (message):
  LIBR_LIB_DIR was not set!
Call Stack (most recent call first):
  cmake/modules/FindLibR.cmake:154 (create_rlib_for_msvc)
  CMakeLists.txt:165 (find_package)
```
Again, I'm new, but I can't figure out how to address this. How do I specify the LIBR_LIB_DIR ?

## Environment info
Operating System: Win10 Pro

Compiler: cmake Visual Studio 2017 win64 with VS 2015 toolcjhains

Package used (python/R/jvm/C++): R

`xgboost` version used: last on GitHub

If installing from source, please provide 

1. The commit hash (`git rev-parse HEAD`): 98be9aef9aab76ca8e1ca6c250c565dcdbcb7be0
Log: 
[CMakeOutput.log](https://github.com/dmlc/xgboost/files/1673413/CMakeOutput.log)

If you are using R package, please provide

1. The R `sessionInfo()`
```
R version 3.4.3 (2017-11-30)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252 
[2] LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets 
[6] methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.4.3 tools_3.4.3   
```

## Steps to reproduce

1. I am trying to build xgboost R package with GPU support as listed [here](http://xgboost.readthedocs.io/en/latest/build.html#installing-r-package-with-gpu-support)
2. After installing VS, CUDA & Git, I tried to build the package with `cmake .. -G"Visual Studio 15 2017 Win64" -DUSE_CUDA=ON -DR_LIB=ON`
3. Response with errors are: 
```
CUDA 9.0 detected, adding Volta compute capability (7.0).
'C:\PROGRA~1\R\R-34~1.3/bin/x64/Rterm.exe" --slave --no-save -e "cat' is not recognized as an internal or external command,
operable program or batch file.
-- LIBR_HOME [/registry]
-- LIBR_EXECUTABLE [C:/Program Files/R/R-3.4.3/bin/x64/R.exe]
-- LIBR_INCLUDE_DIRS [/registry/include]
-- LIBR_LIB_DIR [/registry/bin/x64]
-- LIBR_CORE_LIBRARY []
CMake Error at cmake/modules/FindLibR.cmake:44 (message):
  LIBR_LIB_DIR was not set!
Call Stack (most recent call first):
  cmake/modules/FindLibR.cmake:154 (create_rlib_for_msvc)
  CMakeLists.txt:165 (find_package)
```

-- Configuring incomplete, errors occurred!
See also "C:/Users/<user>/Downloads/xgboost/build/CMakeFiles/CMakeOutput.log".

## What have you tried?

1. git submodule init/udpate
2.
3.

**Comment by Laurae2** [comment]
@Hartsord This message is a non GPU one related. Can you try compiling it first without GPU support, but keeping R support?

**Comment by gm209** [comment]
Hey Guys, 

I also get the same error when trying to build from source,

I've tried the following iterations:

1. Building python only (No DR_LIB) with GPU - **This works**
2, Building R with no GPU - **This fails**
3. Building R with GPU - **This also fails :(**

Session Info:

        R version 3.4.3 (2017-11-30)
        Platform: x86_64-w64-mingw32/x64 (64-bit)
        Running under: Windows >= 8 x64 (build 9200)

        Matrix products: default

        locale:
        [1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252                            LC_MONETARY=English_United Kingdom.1252
        [4] LC_NUMERIC=C                            LC_TIME=English_United Kingdom.1252    

        attached base packages:
        [1] stats     graphics  grDevices utils     datasets  methods   base     

        other attached packages:
        [1] bindrcpp_0.2        ggplot2_2.2.1       xgboost_0.6.4.1     dplyr_0.7.4         h2o_3.16.0.3        data.table_1.10.4-3

        loaded via a namespace (and not attached):
         [1] Rcpp_0.12.15     bindr_0.1        magrittr_1.5     munsell_0.4.3    colorspace_1.3-2 lattice_0.20-35  R6_2.2.2        
         [8] rlang_0.1.6      plyr_1.8.4       tools_3.4.3      grid_3.4.3       gtable_0.2.0     lazyeval_0.2.1   assertthat_0.2.0
        [15] tibble_1.4.2     Matrix_1.2-12    bitops_1.0-6     RCurl_1.95-4.10  glue_1.2.0       stringi_1.1.6    compiler_3.4.3  
        [22] pillar_1.1.0     scales_0.5.0     jsonlite_1.5     pkgconfig_2.0.1 `

Code:  

        cmake .. -G "Visual Studio 15 2017 Win64" -DR_LIB=ON 

Output:

       'C:\PROGRA~1\R\R-34~1.3\bin\x64\R.exe" --slave --no-save -e "cat' is not recognized as an internal        or external command,
       operable program or batch file.
       -- LIBR_HOME [/registry]
       -- LIBR_EXECUTABLE [C:/Program Files/R/R-3.4.3/bin/R.exe]
       -- LIBR_INCLUDE_DIRS [/registry/include]
       -- LIBR_LIB_DIR [/registry/bin/x64]
       -- LIBR_CORE_LIBRARY []
       CMake Error at cmake/modules/FindLibR.cmake:44 (message):
         LIBR_LIB_DIR was not set!
       Call Stack (most recent call first):
         cmake/modules/FindLibR.cmake:154 (create_rlib_for_msvc)
         CMakeLists.txt:165 (find_package)

Any ideas?

Cheers

**Comment by khotilov** [comment]
This is odd... and you both seem to have Windows 8...
Which `cmake --version` do you have? 
Does providing ` -DLIBR_EXECUTABLE="C:/Program Files/R/R-3.4.3/bin/x64/R.exe"` to cmake command result in anything different?
Does running `"C:/Program Files/R/R-3.4.3/bin/R.exe" --slave --no-save -e "cat(normalizePath(R.home(), winslash='/'))"` in CMD work?

**Comment by gm209** [comment]
**This is odd... and you both seem to have Windows 8...**

I've got windows 10 - session.info() doesn't seem to have caught up yet.

        OS Name	 Microsoft Windows 10 Pro

**Which cmake --version do you have?**

        cmake --version
        cmake version 3.10.0-rc5

**Does providing -DLIBR_EXECUTABLE="C:/Program Files/R/R-3.4.3/bin/x64/R.exe" to cmake command result in anything different?**

Tried this and got the same error.

**Does running "C:/Program Files/R/R-3.4.3/bin/R.exe" --slave --no-save -e "cat(normalizePath(R.home(), winslash='/'))" in CMD work?**

When i run the command verbatim on the command line i get the same error;

        "C:/Program Files/R/R-3.4.3/bin/R.exe" --slave --no-save -e "cat(normalizePath(R.home(), winslash='/'))"
        'C:\PROGRA~1\R\R-34~1.3\bin\x64\R.exe" --slave --no-save -e "cat' is not recognized as an internal or external command, operable program or batch file.

However when i run the command inside an R session it works fine:

        cat(normalizePath(R.home(), winslash='/'))
        C:/Program Files/R/R-3.4.3

I managed to make the command run from the command line by removing the winslash argument - this makes me think its a quoting/escape character issue. Is there anyway to get around this by manually setting the path? (I tried creating an environment variable called R_HOME but that didn't work either)

        "C:/Program Files/R/R-3.4.3/bin/R.exe" --slave --no-save -e "cat(normalizePath(R.home()))"
        C:\Program Files\R\R-3.4.3

Cheers

**Comment by khotilov** [comment]
The root cause seems to be that the published 3.4.3 R version has issues handling spaces in the argument of -e. Either removing the space or using a different R version, e.g., 3.4.3patched, makes it work right:
```
C:\>"C:\Program Files\R\R-3.4.2\bin\R.exe" --slave --no-save -e "cat(normalizePath(R.home(), winslash='/'))"
C:/Program Files/R/R-3.4.2
C:\>"C:\Program Files\R\R-3.4.3\bin\R.exe" --slave --no-save -e "cat(normalizePath(R.home(), winslash='/'))"
'C:\PROGRA~1\R\R-34~1.3\bin\x64\R.exe" --slave --no-save -e "cat' is not recognized as an internal or external command,
operable program or batch file.

C:\>"C:\Program Files\R\R-3.4.3patched\bin\R.exe" --slave --no-save -e "cat(normalizePath(R.home(), winslash='/'))"
C:/Program Files/R/R-3.4.3patched
C:\>
C:\>"C:\Program Files\R\R-3.4.3\bin\R.exe" --slave --no-save -e "cat(normalizePath(R.home(),winslash='/'))"
C:/Program Files/R/R-3.4.3
```
I don't want to use shortPathName() since 8dot3 names are disabled on some systems.

**Comment by gm209** [comment]
It works! thanks for your help khotilov!

Cheers
